### ISSUE TITLE
registers/vertical_reuse blk-level simulation fails with ghdl/VHDL DUT

### ISSUE THREAD

**Issue by tpoikela** [tpoikela/uvm-python#26]
When running the above-mentioned example, uvm_fatal is thrown. Looks like APB monitor does not register any items during simulation.

Commands to reproduce the issue:
```
cd test/examples/simple/registers/vertical_reuse
SIM=ghdl make TOPLEVEL_LANG=vhdl IMG=BLK SIMARGS="-- +UVM_TESTNAME=blk_R_test" SIM_BUILD=sim_build_blk
```

Expected result:
Test passes, and does not throw `uvm_fatal`.

**Comment by cmarqu** [comment]
Leaving out `TOPLEVEL_LANG=vhdl` and the `--` makes this config pass with Questa but fail with a `ValueError` in Xcelium, and a segfault with VCS.
So I would consider this test a bit "special" already with Verilog, not necessarily with VHDL/ghdl only.

FWIW, cocotb master reports `RuntimeWarning: Timer setup with value 0, which might exhibit undefined behavior in some simulators` in `src/uvm/base/uvm_globals.py:461`, `src/uvm/base/uvm_globals.py:351`, `src/uvm/base/uvm_globals.py:469`, `test/examples/integrated/apb/apb_master.py:152`, `test/examples/integrated/apb/apb_master.py:156`, which may  have something to do with it.

**Comment by tpoikela** [comment]
I saw this RuntimeWarning as well. I don't think it has been there in the previous versions in July/August. apb_master should work without it, but I don't think it will be so easy to remove from uvm_globals. 

I think it has problems, as right now I'm assuming that `await Timer(0)` would be equal to `#0` in SystemVerilog. But this is not really the case. This goes somewhat deep into scheduling and uvm_phases, and understanding how to solve it without `await Timer(0)` is not trivial.

I suppose the Xcelium error can be related to scheduling, if monitor tries to read the value before driver has driven it to known state. I will cross-check the time stamps with iverilog simulation, to see if there's something different.

**Comment by cmarqu** [comment]
Yes, the behavior has not changed, but the warning has been added - which is probably pointing us at the right place now.
A replacement for the `Timer(0)` might be `ReadOnly`.

**Comment by tpoikela** [comment]
It seems `ReadOnly` correspond to PostPoned region in SV. In uvm_globals, we would need to wait until NBA region only. In this region, values are not yet stable and signal updates can still occur before proceeding to next time step. I tried `ReadWrite` instead, and most of the tests pass using Icarus. I need to debug tomorrow the failures, and see if this is the correct solution.

However, I have another task `uvm_zero_delay`, which simply calls `await Timer(0)`. This was added as UVM uses `#0` in several places. I need to check if this task can be also changed to something else.

**Comment by ktbarrett** [comment]
@tpoikela `ReadWrite` is the correct trigger. See my mostly-complete [documentation on cocotb's timing model](https://github.com/cocotb/cocotb/wiki/Timing-Model). `Timer(0)` is not safe and I have seen it crash simulators in certain circumstances.

**Comment by tpoikela** [comment]
@ktbarrett Thanks for this info! The documentation is excellent and very concise. It will be a slightly bigger effort to adapt uvm-python to follow this (unless removing zero delays works out of the box). 

For verilator, I'm only using `ReadWrite` (no `Timer(0)`), but iverilog requires `Timer(0)` in order to make all the tests pass. verilator has problems with some of the tests though.

**Comment by ktbarrett** [comment]
@tpoikela `Timer(0)` has also been used to unconditionally yield control to the scheduler to allow another coroutine to run. In those cases you could replace it with `NullTrigger`.

**Comment by tpoikela** [comment]
@ktbarrett Thanks for this tip. Sounds like NullTrigger might work here.

**Comment by tpoikela** [comment]
@themperek Does this mean they need to be inside the same coroutine, RisingEdge followed by ReadWrite as next trigger?

@ktbarrett  I've managed to remove `Timer(0)`, and replaced it with `NullTrigger`. This seems to have improved verilator stability as well, and 2 testcases which were failing for verilator are now passing. Is `NullTrigger` something which does not interact with simulator at all since it's used for internal scheduling?

**Comment by ktbarrett** [comment]
>Is NullTrigger something which does not interact with simulator at all since it's used for internal scheduling?

That's correct, it does not interact with the simulator, it's a Python-only trigger. Glad to hear it fixed some things.

**Comment by tpoikela** [comment]
Closing this as very old issue.
